cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -443,11 +452,27 @@ mkConsensusTracers trSel verb tr nodeKern fStats = do
   blockForgeOutcomeExtractor <- mkOutcomeExtractor
   elidedFetchDecision <- newstate  -- for eliding messages in FetchDecision tr
   forgeTracers <- mkForgeTracers
+  meta <- mkLOMeta Critical Public
+
+  let trmet = appendName "metrics" tr
+
+  headerMessageCount <- STM.newTVarIO 0
+  blockMessageCount <- STM.newTVarIO 0
 
   pure Consensus.Tracers
     { Consensus.chainSyncClientTracer = tracerOnOff (traceChainSyncClient trSel) verb "ChainSyncClient" tr
-    , Consensus.chainSyncServerHeaderTracer = tracerOnOff (traceChainSyncHeaderServer trSel) verb "ChainSyncHeaderServer" tr
-    , Consensus.chainSyncServerBlockTracer = tracerOnOff (traceChainSyncHeaderServer trSel) verb "ChainSyncBlockServer" tr
+    , Consensus.chainSyncServerHeaderTracer = tracerOnOff' (traceChainSyncHeaderServer trSel) $
+        Tracer $ \ev -> do
+          traceWith (annotateSeverity . toLogObject' verb $ appendName "ChainSyncHeaderServer" tr) ev
+          when (isRollForward ev) $ do
+            count <- STM.atomically $ STM.modifyTVar headerMessageCount (+1) >> STM.readTVar headerMessageCount
+            traceI trmet meta "HeaderRollForward" count
```

Comment:
I'm not satisfied with how this is done.  It seems quite imperative.  Any suggestions?